domain/upgrade/state/state.go

```diff
@@ -200,72 +201,65 @@ func (st *State) SetControllerDone(ctx context.Context, upgradeUUID, controllerI
 	lookForDoneNodesQuery := `
 SELECT (controller_node_id, node_upgrade_completed_at) AS &infoControllerNode.*
 FROM   upgrade_info_controller_node
-WHERE  upgrade_info_uuid = $M.info_uuid`
-	lookForDoneNodesStatement, err := sqlair.Prepare(lookForDoneNodesQuery, infoControllerNode{}, sqlair.M{})
+WHERE  upgrade_info_uuid = $M.info_uuid
+  AND controller_node_id = $M.controller_id;`
+	lookForDoneNodesStmt, err := sqlair.Prepare(lookForDoneNodesQuery, infoControllerNode{}, sqlair.M{})
 	if err != nil {
-		return errors.Annotatef(err, "preparing %q", lookForDoneNodesQuery)
+		return errors.Annotatef(err, "preparing select done query")
 	}
 
 	setNodeToDoneQuery := `
 UPDATE upgrade_info_controller_node
 SET    node_upgrade_completed_at = DATETIME("now")
-WHERE  upgrade_info_uuid = $M.info_uuid 
-       AND controller_node_id = $M.controller_id`
-	setNodeToDoneStatement, err := sqlair.Prepare(setNodeToDoneQuery, sqlair.M{})
+WHERE  upgrade_info_uuid = $M.info_uuid
+       AND controller_node_id = $M.controller_id
+	   AND node_upgrade_completed_at IS NULL;
+`
+	setNodeToDoneStmt, err := sqlair.Prepare(setNodeToDoneQuery, sqlair.M{})
 	if err != nil {
-		return errors.Annotatef(err, "preparing %q", setNodeToDoneQuery)
+		return errors.Annotatef(err, "preparing update node query")
 	}
 
 	completeUpgradeQuery := `
 UPDATE upgrade_info
 SET    completed_at = DATETIME("now")
-WHERE  uuid = $M.info_uuid`
-	completeUpgradeStatement, err := sqlair.Prepare(completeUpgradeQuery, sqlair.M{})
+WHERE  uuid = $M.info_uuid
+AND (
```

Comment:
I think you could do this more cogently with a single `WITH` at the top that uses a `HAVING` clause.

Reply:
I did try with `WITH` but I hit a road block. I can look again another time.